# 4.8 Grouping Data and Aggregating Variables 

1. Import libraries and data set 
2. Aggregating data with agg() function
3. Creating loyalty flag with transform() and loc()
4. Looking at spending habits of different customers
5. Creating a spending flag
6. Creating an order frequency flag 

In [1]:
# import libraties
import pandas as pd
import numpy as np
import os

In [2]:
# Path
path = r"C:\Users\Natalia\Documents\Instacart Basket Analysis"

In [3]:
#import ords_prods_merge dataset
df = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'full_merged4.pkl'))

In [4]:
#check dataframe output
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

     busiest_days busiest_period_of_day  max_order  loyalty_flag  
0  Regularly busy        Average orders         10  New customer  
1    Slowest days        Average orders         10  New customer  
2    Slowest days           Most orders         10  New customer  
3    Slowest days        Average orders         10  New customer  
4    Slowest days           Most orders         10  New customer

In [5]:
#drop max_order and loyalty flags columns since I will be recreating them as part of this submission
df = df.drop(['max_order', 'loyalty_flag'], axis = 1)

In [6]:
#checking out put 
print('Shape of new dataframe:', df.shape)

Shape of new dataframe: (32404859, 17)


In [7]:
print('Sample of new dataframe')
df.sample(5)

Sample of new dataframe


order_id  user_id  order_number  orders_day_of_week  \
2806554      21414   137517            77                   3   
19409841   1605039    60616            40                   0   
1690676    1545298    70298            17                   1   
25996244   2816073    70235            18                   0   
31675090   1276966    48412             9                   6   

          order_hour_of_day  days_since_prior_order  product_id  \
2806554                  11                     2.0       39877   
19409841                 15                     5.0       29615   
1690676                  12                    11.0       24852   
25996244                 10                     7.0       29013   
31675090                  9                    14.0       16939   

          add_to_cart_order  reordered _merge  \
2806554                   2          1   both   
19409841                  3          1   both   
1690676                   7          1   both   
25996244                  4          1   both   
31675090                  1          1   both   

                                          product_name  aisle_id  \
2806554                     Organic Granny Smith Apple        24   
19409841                     Bread, Country Buttermilk       112   
1690676                                         Banana        24   
25996244  Gluten Free Cracked Pepper And Salt Crackers        78   
31675090           Milk, Lowfat, Chocolate, 1% Milkfat        84   

          department_id  prices    price_range_loc    busiest_days  \
2806554               4     1.3  Low_range product    Slowest days   
19409841              3     2.8  Low_range product    Busiest days   
1690676               4    12.3  Mid-range product    Busiest days   
25996244             19     6.6  Mid-range product    Busiest days   
31675090             16     5.4  Mid-range product  Regularly busy   

         busiest_period_of_day  
2806554            Most orders  
19409841           Most orders  
1690676            Most orders  
25996244           Most orders  
31675090           Most orders

## 1. Aggregating data with the agg() function

#### Aggregate the mean of the "order_number" column grouped by "department_id" for the entire dataframe

In [8]:
# group df_ords_prods_merge by the 'department_id' column, then calculate the mean, min and max of the 'order_number' column
print('The mean, min and max of order_number grouped by the department_id column in df:')
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

The mean, min and max of order_number grouped by the department_id column in df:


order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

#### Creating a subset to compare aggregations against the entire dataframe

In [9]:
#creating subset and printing to check output
df_subset = df[:1000000]

print('First five results of df_ords_prods_merge subset')
df_subset.head(5)

First five results of df_ords_prods_merge subset


order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

     busiest_days busiest_period_of_day  
0  Regularly busy        Average orders  
1    Slowest days        Average orders  
2    Slowest days           Most orders  
3    Slowest days        Average orders  
4    Slowest days           Most orders

In [10]:
# group df_subset by the 'department_id' column, then calculate the mean, min and max of the 'order_number' column
print('The mean, min and max of order_number grouped by the department_id column in df_subset:')
df_subset.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

The mean, min and max of order_number grouped by the department_id column in df_subset:


order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

Observations: whilst there are some difference between the aggregations seen in the subset compared to the entire dataframe, they tend to be within reasonable ranges. I calculated the percentage change for four departments: department 1 was within 4%, department 8 was within 7%, department was within 2%, and department 21 was within 4%. 

There is some variance in the max values in the subset. The max values for all departments in the entire dataset is 99, however, in some of the subset department it is lower, for example, department 8 is only 91. Again, this is not unreasonable to expect when only looking at a slice of the data. 

## 2. Creating a loyalty flag for customers using transform() and loc() functions

#### Create a new column, 'max_order' to place the results of the maximum orders for each user

In [11]:
# create a new column 'max_order2' containing the maximum order number for each user_id, calculated using the groupby() and transform() functions
df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
#check output by looks at user_id and max_order
print('Sample user_id and max_order output:')
df[['user_id', 'max_order']].sample(10)

Sample user_id and max_order output:


user_id  max_order
7404896    140130         55
5328502     81601          7
2308376     26406         37
25912367   172053         37
2664731     42151         24
20059923   132161         41
22700068     7581         27
13432077   199074         26
29645416   174302         25
13768328   103490         14

In [13]:
df.sample(5)

order_id  user_id  order_number  orders_day_of_week  \
3105678    1630877    20558            61                   2   
643454     2031077    38263             6                   1   
9361212    1315288    29932             4                   6   
11927976   1913658    79238            69                   1   
7444954    1496358    59128            10                   2   

          order_hour_of_day  days_since_prior_order  product_id  \
3105678                  13                     3.0       21903   
643454                    0                    29.0       35951   
9361212                  12                     1.0        9477   
11927976                 10                     0.0       20119   
7444954                  15                     6.0       44359   

          add_to_cart_order  reordered _merge  \
3105678                  12          1   both   
643454                    2          1   both   
9361212                   1          0   both   
11927976                  1          1   both   
7444954                  21          1   both   

                             product_name  aisle_id  department_id  prices  \
3105678              Organic Baby Spinach       123              4     8.2   
643454    Organic Unsweetened Almond Milk        91             16     1.3   
9361212             Medium Cheddar Cheese        21             16    11.5   
11927976            Sparkling Water Berry       115              7    13.4   
7444954        Organic Small Bunch Celery        83              4     8.4   

            price_range_loc    busiest_days busiest_period_of_day  max_order  
3105678   Mid-range product  Regularly busy           Most orders         99  
643454    Low_range product    Busiest days         Fewest orders          8  
9361212   Mid-range product  Regularly busy           Most orders         63  
11927976  Mid-range product    Busiest days           Most orders         81  
7444954   Mid-range product  Regularly busy           Most orders         10

#### Derive a new column, 'loyalty_flag' using the loc() function and the values found in "max_order"

In [14]:
df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [15]:
df.loc[df['max_order'] <= 40, 'loyalty_flag'] = 'Regular customer'

In [16]:
df.loc[df['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [17]:
#checking ouput of flag creation
df['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [18]:
# checking the accuracy of the 'loyalty_flag' column with output from columns 'user_id', 'max_order', and 'loyalty_flag' column
print('Sample output from columns user_id, max_order and loyalty_flag columns:')
df[['user_id', 'max_order', 'loyalty_flag']].sample(20)

Sample output from columns user_id, max_order and loyalty_flag columns:


user_id  max_order      loyalty_flag
11665510    31161         50    Loyal customer
4609941    102655         69    Loyal customer
6083957    128340          8      New customer
21835767    85453         16  Regular customer
4684073     69315         21  Regular customer
442944     205972         99    Loyal customer
3264157    155301         52    Loyal customer
16340992    78326         15  Regular customer
23993250    85040         10      New customer
14809822    28132         48    Loyal customer
22936920    42835         45    Loyal customer
9698050     61055         19  Regular customer
589015     140221         34  Regular customer
5835769    110127          5      New customer
9472924    154807         41    Loyal customer
9165121     85954         26  Regular customer
10072618   107456         26  Regular customer
3937350    131606         26  Regular customer
3764457    179727         40  Regular customer
19746835   102033         22  Regular customer

## 3. Looking at the spending habits of different customers

#### Check basic statistics of product prices for each loyalty category (loyal, regular, and new). Do the prices of products purchased by each loyalty group differ?

In [19]:
#basic statistical comparison of 'loyalty_flag' column, followed by calculation of mean, min, max, and sum of the 'prices' column
print('The mean, min, max, and sum of prices grouped by the loyalty_flag:')

df.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'sum']})

The mean, min, max, and sum of prices grouped by the loyalty_flag:


prices                           
                       mean  min      max          sum
loyalty_flag                                          
Loyal customer    10.386336  1.0  99999.0  106814042.2
New customer      13.294670  1.0  99999.0   83011787.2
Regular customer  12.495717  1.0  99999.0  198391693.2

In [20]:
#confirming the numbers of each type of customer
df.groupby('loyalty_flag')['user_id'].count()

loyalty_flag
Loyal customer      10284093
New customer         6243990
Regular customer    15876776
Name: user_id, dtype: int64

##### Calculating each loyalty flag as a percentage of total orders

In [21]:
#setting loyalty_flag as variable 
loyalty_flag = df.loyalty_flag

In [22]:
#calculating loyalty_flag as % of whole
percent_loyalty_flag = loyalty_flag.value_counts(normalize = True)
print(percent_loyalty_flag)

Regular customer    0.489950
Loyal customer      0.317363
New customer        0.192687
Name: loyalty_flag, dtype: float64


In [23]:
#returning results in easier to read format 
percent_loyalty_flag_100 = loyalty_flag.value_counts(normalize = True).mul(100).round(1).astype(str) + '%'
print(percent_loyalty_flag_100)

Regular customer    49.0%
Loyal customer      31.7%
New customer        19.3%
Name: loyalty_flag, dtype: object


### Observations

Almost 50% of orders fall within the regular customer category, just over 30% are from loyal customers, and just under 20% are new customers. Interestingly, the average price for loyal customers is the lowest of the three groups whilst new customers have the highest which I was surprised by. Some interesting futher analysis could be focused on the types of products loyal customers are buying.    

The maximum price value of $99,999 seems like an extreme outlier, and requires further investigation. Significant outliers like this could skew our data.  

##### Investigating the outlier prices

In [24]:
#returning a list of all prices, ordered by most expensive
df['prices'].drop_duplicates().sort_values(ascending = False)

29165516    99999.0
10030345    14900.0
18504754       25.0
25758883       24.9
28425432       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

There are two prices which appear to be outliers: 99,999 and 14,900

In [25]:
#filtering the dataset by two outlier prices
max_price = df.loc[df['prices'].isin([99999, 14900])]

In [26]:
max_price.sample(5)

order_id  user_id  order_number  orders_day_of_week  \
10034147   2808820   177179            59                   3   
10032616   1260580   105739            64                   2   
29165556   1356767    14872            13                   4   
29165953   1627230   142223            41                   2   
10031962   2307146    79117             6                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
10034147                 12                     3.0       21553   
10032616                 14                     4.0       21553   
29165556                 14                     5.0       33664   
29165953                 14                     2.0       33664   
10031962                  8                     5.0       21553   

          add_to_cart_order  reordered _merge  \
10034147                  6          1   both   
10032616                  2          1   both   
29165556                 29          0   both   
29165953                 17          1   both   
10031962                  2          1   both   

                              product_name  aisle_id  department_id   prices  \
10034147  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10032616  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
29165556             2 % Reduced Fat  Milk        84             16  99999.0   
29165953             2 % Reduced Fat  Milk        84             16  99999.0   
10031962  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   

             price_range_loc    busiest_days busiest_period_of_day  max_order  \
10034147  High-range product    Slowest days           Most orders         94   
10032616  High-range product  Regularly busy           Most orders         72   
29165556  High-range product    Slowest days           Most orders         18   
29165953  High-range product  Regularly busy           Most orders         53   
10031962  High-range product    Busiest days        Average orders         12   

              loyalty_flag  
10034147    Loyal customer  
10032616    Loyal customer  
29165556  Regular customer  
29165953    Loyal customer  
10031962  Regular customer

In [27]:
max_price.shape

(5127, 19)

There are 5127 rows impacted by these outlier prices of 99,999 and  14,900. Ideally I would conduct some more investigation to see if we can figure out the correct price and amend the dataset to reflect. Results including these values as they currently are may impact the mean values returned in the basic statistics returned above. 

## 3. Create a spending flag for each user based on the average price across all their order

* If the mean of the prices of products purchased by a user is lower than 10, then flag them as a "Low spender".
* If the mean of the prices of products purchased by a user is higher than or equsl to 10, then flag them as a "High spender"

In [28]:
#create a new columns called "spending_flag"
df['avg_price'] = df.groupby(['user_id'])['prices'].transform(np.mean)

In [29]:
#round the averages
df['avg_price'] = df.groupby(['user_id'])['prices'].transform(np.mean).round(2)

In [30]:
#check output
print('Sample of user_id and average price output:')
df[['user_id', 'avg_price']].sample(10)

Sample of user_id and average price output:


user_id  avg_price
25339016   182153       7.77
27858882    24503       6.39
6049269     22909       7.61
6705923    151443       7.35
5991926    104948       8.35
11539615    79095       5.56
29149293   183824       8.04
29038193    33464       9.06
26791911   123062       7.26
5029909     66155       8.18

#### Create a flag which tags any user with an average spend of more than or equal to 10, as "high spender". Any customers with an average spend of less than 10 will be marked as "Low spender"

In [31]:
df.loc[df['avg_price'] >= 10, 'spending_flag']='High spender'

In [32]:
df.loc[df['avg_price'] < 10, 'spending_flag']='Low spender'

In [33]:
#check output
df.sample(5)

order_id  user_id  order_number  orders_day_of_week  \
26370114   1921362   197735             2                   6   
30125791   2070655    65211             4                   5   
3095467     144296    11914            29                   4   
8422498    1718195    61216            14                   2   
26987633    133666    17660             1                   5   

          order_hour_of_day  days_since_prior_order  product_id  \
26370114                 17                     6.0       17207   
30125791                 19                     7.0       39815   
3095467                  15                     6.0       21903   
8422498                  10                    30.0       44987   
26987633                 14                     NaN       35761   

          add_to_cart_order  reordered _merge  ... aisle_id  department_id  \
26370114                  3          0   both  ...      120             16   
30125791                  5          0   both  ...       94              7   
3095467                   3          1   both  ...      123              4   
8422498                  12          1   both  ...       53             16   
26987633                 15          0   both  ...       21             16   

          prices    price_range_loc    busiest_days busiest_period_of_day  \
26370114     2.6  Low_range product  Regularly busy        Average orders   
30125791     1.7  Low_range product  Regularly busy        Average orders   
3095467      8.2  Mid-range product    Slowest days           Most orders   
8422498      9.1  Mid-range product  Regularly busy           Most orders   
26987633    13.3  Mid-range product  Regularly busy           Most orders   

         max_order      loyalty_flag avg_price  spending_flag  
26370114        23  Regular customer      8.69    Low spender  
30125791        12  Regular customer      7.19    Low spender  
3095467         60    Loyal customer      7.19    Low spender  
8422498         31  Regular customer      7.68    Low spender  
26987633         3      New customer      9.94    Low spender  

[5 rows x 21 columns]

## 4. Create an order frequency flag which corresponds to the regularity of a customer's order behaviour

* If the median of "days_since_prior_order" is **higher than 20**, the customer should be labeled a **"Non-frequent customer"**
* If the median of "days_since_prior_order" is **higher than 10 and lower than or equal to 20**, then the customer should be labeled a **"Regular customer"**
* If the median is **lower than or equal to 10**, then the customer should be labeled a **"Frequent customer"**

In [34]:
#create a new column "order_freq" containing the median in days_since_prior_order
df['order_freq'] = df.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [35]:
# check the output of 'user_id' and 'order_freq'
print('Sample user_id and order_freq output:')
df[['user_id', 'order_freq']].sample(10)

Sample user_id and order_freq output:


user_id  order_freq
16463173    85954         3.0
29570846    89209        13.0
18446404   201774         7.0
28267899   131805         2.0
1023608     14240         5.0
24153021   185218        13.0
21004090    62048         8.0
17389515   104079        21.0
26496895    53744         4.0
24605535   166698        30.0

#### Create a flag which tags any user with a median frequency order of less than or equal to 10 as "non-frequent customer", higher than 10 and lower than or equal to 20 as "regular customer", and higher than 20 as "frequent customer"

In [36]:
# create the 'order_freq_flag' column based on each user's median in days_since_prior_order
df.loc[df['order_freq'] > 20, 'order_freq_flag'] = 'Non-frequent customer'

In [37]:
df.loc[(df['order_freq'] > 10) & (df['order_freq'] <= 20), 'order_freq_flag'] = 'Regular customer'

In [38]:
df.loc[df['order_freq'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [39]:
#check the output and count of each new flag

print('Count of each order frequency variable')
df['order_freq_flag'].value_counts(dropna = False)

Count of each order frequency variable


Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_freq_flag, dtype: int64

# Export Dataset

In [40]:
df.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'full_merged5.pkl'))